### PR TITLE
Balance Checker and BalanceAPI UML

### PR DESCRIPTION
**Pull request by gasiortomasz**

### CODE REVIEW

File: universal-login-sdk/lib/integration/ethereum/BalanceChecker.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import {providers, Contract} from 'ethers';
+import IERC20 from 'openzeppelin-solidity/build/contracts/IERC20.json';
+import {ETHER_NATIVE_TOKEN} from '@universal-login/commons';
+
+export class BalanceChecker {
+  constructor(private provider: providers.Provider, private walletAddress: string) {}
+
+  getBalance(tokenAddress: string) {
```
Comment (JustynaBroniszewska) [inline universal-login-sdk/lib/integration/ethereum/BalanceChecker.ts]: code duplication? [balance.ts](https://github.com/UniversalLogin/UniversalLoginSDK/blob/master/universal-login-commons/lib/integration/ethereum/balance.ts)

File: universal-login-sdk/test/integration/integration/BalanceChecker.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    it('1 token', async () => {
+      await mockToken.transfer(TEST_ACCOUNT_ADDRESS, utils.bigNumberify('1'));
+      const balance = await balanceChecker.getBalance(mockToken.address);
+      expect(balance).to.eq(utils.bigNumberify('1'));
+    });
+
+    it('not deployed', async () => {
+      await mockToken.transfer(TEST_ACCOUNT_ADDRESS, utils.bigNumberify('1'));
+      await expect(balanceChecker.getBalance('0x000000000000000000000000000000000000DEAD'))
+        .to.be.rejected;
```
Comment (marekkirejczyk) [inline universal-login-sdk/test/integration/integration/BalanceChecker.ts]: rejectedWith

File: universal-login-sdk/lib/integration/ethereum/BalanceChecker.ts (hunk 1 above)
Comment (marekkirejczyk) [reply universal-login-sdk/lib/integration/ethereum/BalanceChecker.ts]: Yes, we should deduplicate in the same PR.

File: universal-login-sdk/lib/integration/ethereum/BalanceChecker.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+  private getEthBalance() {
+    return this.provider.getBalance(this.walletAddress);
+  }
+
+  private getTokenBalance(tokenAddress: string) {
+    const token = new Contract(tokenAddress, IERC20.abi, this.provider);
+    return token.balanceOf(this.walletAddress);
+  }
+}
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/integration/ethereum/BalanceChecker.ts]: Shall we move it to commons

File: universal-login-commons/test/integration/ethereum/RequiredBalanceChecker.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  const supportedTokens = [{address: ETHER_NATIVE_TOKEN.address, minimalAmount: utils.parseEther('0.5').toString()}];
+
+  it('no tokens with required balance', async () => {
+    expect(await requiredBalanceChecker.findTokenWithRequiredBalance(supportedTokens, TEST_ACCOUNT_ADDRESS)).to.be.null;
+  });
+
+  it('token with just enough balance', async () => {
+    await wallet.sendTransaction({to: TEST_ACCOUNT_ADDRESS, value: utils.parseEther('0.5')});
+    expect(await requiredBalanceChecker.findTokenWithRequiredBalance(supportedTokens, TEST_ACCOUNT_ADDRESS)).to.eq(ETHER_NATIVE_TOKEN.address);
+  });
```
Comment (marekkirejczyk) [inline universal-login-commons/test/integration/ethereum/RequiredBalanceChecker.ts]: Would be nice to have test cases with two items (zero, one, two balances meeting conditions).

File: universal-login-relayer/lib/integration/ethereum/WalletService.ts [hunk 5]
```diff
@@ -11,11 +11,13 @@ import {WalletDeployer} from '../ethereum/WalletDeployer';
 class WalletService {
   private bytecode: string;
   private abi: Abi;
+  private requiredBalanceChecker: RequiredBalanceChecker;
 
   constructor(private wallet: Wallet, private config: Config, private ensService: ENSService, private hooks: EventEmitter, private walletDeployer: WalletDeployer) {
     const contractJSON = ProxyContract;
     this.abi = contractJSON.interface;
     this.bytecode = `0x${contractJSON.evm.bytecode.object}`;
+    this.requiredBalanceChecker = new RequiredBalanceChecker(new BalanceChecker(this.wallet.provider));
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/integration/ethereum/WalletService.ts]: Should be injected.

File: universal-login-sdk/lib/core/observers/BalanceObserver.ts [hunk 6]
```diff
@@ -8,9 +8,11 @@ export type ReadyToDeployCallback = (tokenAddress: string, contractAddress: stri
 export class BalanceObserver extends ObserverRunner {
   private contractAddress?: string;
   private callback?: ReadyToDeployCallback;
+  private requiredBalanceChecker: RequiredBalanceChecker;
 
   constructor(private supportedTokens: SupportedToken[], private provider: providers.Provider) {
     super();
+    this.requiredBalanceChecker = new RequiredBalanceChecker(new BalanceChecker(this.provider));
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/core/observers/BalanceObserver.ts]: Should be injected.

**Comment by marekkirejczyk** [comment]
One comment left + need rebase

**Pull request merged**